### PR TITLE
Output pin config

### PR DESCRIPTION
**Pull request by chuckwagoncomputing**
I have made it so OutputPin manages its own pins when the configuration changes like I did with ButtonDebounce. I haven't gone through the code yet and removed individual markUnused code, but I thought I would submit this for review in the mean time.

### CODE REVIEW

File: firmware/controllers/system/efi_gpio.cpp [hunk 1]
```diff
@@ -79,8 +79,10 @@ EnginePins::EnginePins() {
 #if EFI_PROD_CODE
 #define setPinValue(outputPin, electricalValue, logicValue)                        \
   {                                                                                \
-    if ((outputPin)->currentLogicValue != (logicValue)) {                          \
-	  palWritePad((outputPin)->port, (outputPin)->pin, (electricalValue));         \
+    if ((outputPin)->currentLogicValue != (logicValue)) {	\
+	  brain_pin_e tmppin = (outputPin)->brainPin;		\
+	  const char* tmpmsg = (outputPin)->m_msg;		\
```
Comment (rusefillc) [inline firmware/controllers/system/efi_gpio.cpp:84]: I have _zero_ clue what's going on here. Like _no_ clue whatsoever. That's a red flag.

File: firmware/controllers/system/efi_gpio.h [hunk 2]
```diff
@@ -42,11 +42,12 @@ class OutputPin {
 	 * outputMode being a pointer allow us to change configuration (for example invert logical pin) in configuration and get resuts applied
 	 * away, or at least I hope that's why
 	 */
-	void initPin(const char *msg, brain_pin_e brainPin, const pin_output_mode_e *outputMode);
+	void initPin(const char *msg, brain_pin_e &brainPin, const pin_output_mode_e *outputMode);
+	void initPinR(const char *msg, brain_pin_e brainPin, const pin_output_mode_e *outputMode);
```
Comment (rusefillc) [inline firmware/controllers/system/efi_gpio.h:46]: I have no clue what is "R". This whole PR is extremely cryptic :(

File: firmware/controllers/system/efi_gpio.h [hunk 3]
```diff
@@ -61,25 +62,31 @@ class OutputPin {
...
-	ioportid_t port = 0;
-	uint8_t pin = 0;
 	#if (BOARD_EXT_GPIOCHIPS > 0)
 		/* used for external pins */
-		brain_pin_e brainPin;
 		bool ext;
 	#endif
 #endif /* EFI_GPIO_HARDWARE */
+	pin_output_mode_e outputMode;
+	brain_pin_e brainPin;
```
Comment (rusefillc) [inline firmware/controllers/system/efi_gpio.h:71]: field;
fieldPtr;

that's something really really unusual.

overall I am extremely confused by this PR. without adequate comments I see enormous amount of code complexity and risk, there is no way such complexity and risk justifies the benefits of this PR.

**Review by rusefillc — CHANGES_REQUESTED** [review]
I am extremely uncomfortable with how complex this change is and how poorly the ideas are documented.

File: firmware/controllers/system/efi_gpio.cpp (hunk 1 above)
Comment (chuckwagoncomputing) [reply firmware/controllers/system/efi_gpio.cpp:84]: To be fair, I barely changed anything.

File: firmware/controllers/system/efi_gpio.h (hunk 3 above)
Comment (chuckwagoncomputing) [reply firmware/controllers/system/efi_gpio.h:71]: same thing going on here as with ButtonDebounce...

**Comment by chuckwagoncomputing** [comment]
More comments incoming.
I created this PR to get some feedback on how I have done things as a whole. There's a log of little things to do before this is finished.

**Comment by chuckwagoncomputing** [comment]
I have now removed cases where pin changes were handled manually. If you notice any I missed, let me know.

**Comment by rusefillc** [comment]
this is a very early draft PR which was not marked as DRAFT. GH has special considerations for drafts which help to reduce confusion.

please allow me to close this PR and let's reopen it once it's more mature. I believe that at the moment we should really switch focus to #1801

**Pull request closed without merging**